python/ql/test/experimental/dataflow/match/test.py

```diff
@@ -0,0 +1,152 @@
+
+import sys
+import os
+
+sys.path.append(os.path.dirname(os.path.dirname((__file__))))
+from testlib import *
+
+# These are defined so that we can evaluate the test code.
+NONSOURCE = "not a source"
+SOURCE = "source"
+
+
+def is_source(x):
+    return x == "source" or x == b"source" or x == 42 or x == 42.0 or x == 42j
+
+
+def SINK(x):
+    if is_source(x):
+        print("OK")
+    else:
+        print("Unexpected flow", x)
+
+
+def SINK_F(x):
+    if is_source(x):
+        print("Unexpected flow", x)
+    else:
+        print("OK")
+
+def test_guard():
+    match SOURCE:
+        case x if SINK(x): #$ flow="SOURCE, l:-1 -> x"
+            pass
+
+@expects(2)
+def test_as_pattern():
+    match SOURCE:
+        case x as y:
+            SINK(x) #$ flow="SOURCE, l:-2 -> x"
+            SINK(y) #$ flow="SOURCE, l:-3 -> y"
+
+def test_or_pattern():
+    match SOURCE:
+        # We cannot use NONSOURCE in place of "" below, since it would be seen as a variable.
+        case ("" as x) | x:
+            SINK(x) #$ flow="SOURCE, l:-3 -> x"
+
+# No flow for literal pattern
+
+def test_capture_pattern():
+    match SOURCE:
+        case x:
+            SINK(x) #$ flow="SOURCE, l:-2 -> x"
+
+# No flow for wildcard pattern
+
+class Unsafe:
+    VALUE = SOURCE
+
+def test_value_pattern():
+    match SOURCE:
+        case Unsafe.VALUE as x:
+            SINK(x) #$ flow="SOURCE, l:-2 -> x" MISSING: flow="SOURCE, l:-5 -> x"
```

Comment:
This missing flow is curious. Do you know what causes it?

Reply:
Yes. I thought we would need the guard aspect to be implemented, but now that you mention it, there is already flow from the value pattern `Unsafe.VALUE` to the alias `x`. So all we should need is flow from the actual value to the value pattern.

Reply:
Actually, I think we are also missing flow from the SSA variable on line 58 to the attribute at line 62.